File: Tests/ClaudeUsageWidgetCoreTests/ServiceStatusTests.swift
import Foundation
import Testing
@testable import ClaudeUsageWidgetCore

@Suite("ServiceStatus")
struct ServiceStatusTests {
    @Test("component statuses map onto the widget's vocabulary")
    func mapsComponentStatuses() {
        #expect(ServiceStatus.component("operational") == .operational)
        #expect(ServiceStatus.component("degraded_performance") == .degraded)
        #expect(ServiceStatus.component("partial_outage") == .partialOutage)
        #expect(ServiceStatus.component("major_outage") == .majorOutage)
        #expect(ServiceStatus.component("under_maintenance") == .maintenance)
    }

    @Test("page indicators map onto the same vocabulary")
    func mapsIndicators() {
        #expect(ServiceStatus.indicator("none") == .operational)
        #expect(ServiceStatus.indicator("minor") == .degraded)
        #expect(ServiceStatus.indicator("major") == .partialOutage)
        #expect(ServiceStatus.indicator("critical") == .majorOutage)
        #expect(ServiceStatus.indicator("maintenance") == .maintenance)
    }

    @Test("an unrecognised value is unknown, not a guess")
    func unknownValues() {
        #expect(ServiceStatus.component("teleported") == .unknown)
        #expect(ServiceStatus.indicator("") == .unknown)
    }

    @Test("every case has a short label that fits the dial")
    func labels() {
        #expect(ServiceStatus.operational.label == "OK")
        #expect(ServiceStatus.degraded.label == "SLOW")
        #expect(ServiceStatus.partialOutage.label == "PARTIAL")
        #expect(ServiceStatus.majorOutage.label == "DOWN")
        #expect(ServiceStatus.maintenance.label == "MAINT")
        #expect(ServiceStatus.unknown.label == "—")
    }
}

@Suite("StatusDecoder")
struct StatusDecoderTests {
    static let payload = Data("""
    {
      "status": { "indicator": "none", "description": "All Systems Operational" },
      "components": [
        { "name": "claude.ai", "status": "operational" },
        { "name": "Claude Code", "status": "degraded_performance" },
        { "name": "Claude Cowork", "status": "operational" }
      ]
    }
    """.utf8)

    @Test("the Claude Code component wins over the page roll-up")
    func prefersClaudeCode() throws {
        #expect(try StatusDecoder.status(from: Self.payload) == .degraded)
    }

    @Test("the page indicator is the fallback when the component is absent")
    func fallsBackToIndicator() throws {
        let data = Data("""
        {
          "status": { "indicator": "critical" },
          "components": [ { "name": "claude.ai", "status": "operational" } ]
        }
        """.utf8)
        #expect(try StatusDecoder.status(from: data) == .majorOutage)
    }

    @Test("a body with neither component nor indicator is malformed")
    func rejectsUselessBody() {
        #expect(throws: UsageError.malformedResponse) {
            try StatusDecoder.status(from: Data(#"{"components": []}"#.utf8))
        }
    }

    @Test("a non-JSON body is malformed")
    func rejectsGarbage() {
        #expect(throws: UsageError.malformedResponse) {
            try StatusDecoder.status(from: Data("<html>nope</html>".utf8))
        }
    }
}

File: Tests/ClaudeUsageWidget.Core.Tests/ServiceStatusTests.cs
namespace ClaudeUsageWidget.Core.Tests;

public class ServiceStatusTests
{
    [Fact]
    public void MapsComponentStatuses()
    {
        Assert.Equal(ServiceStatus.Operational, ServiceStatusParser.Component("operational"));
        Assert.Equal(ServiceStatus.Degraded, ServiceStatusParser.Component("degraded_performance"));
        Assert.Equal(ServiceStatus.PartialOutage, ServiceStatusParser.Component("partial_outage"));
        Assert.Equal(ServiceStatus.MajorOutage, ServiceStatusParser.Component("major_outage"));
        Assert.Equal(ServiceStatus.Maintenance, ServiceStatusParser.Component("under_maintenance"));
    }

    [Fact]
    public void MapsIndicators()
    {
        Assert.Equal(ServiceStatus.Operational, ServiceStatusParser.Indicator("none"));
        Assert.Equal(ServiceStatus.Degraded, ServiceStatusParser.Indicator("minor"));
        Assert.Equal(ServiceStatus.PartialOutage, ServiceStatusParser.Indicator("major"));
        Assert.Equal(ServiceStatus.MajorOutage, ServiceStatusParser.Indicator("critical"));
        Assert.Equal(ServiceStatus.Maintenance, ServiceStatusParser.Indicator("maintenance"));
    }

    [Fact]
    public void UnknownValuesAreUnknownNotAGuess()
    {
        Assert.Equal(ServiceStatus.Unknown, ServiceStatusParser.Component("teleported"));
        Assert.Equal(ServiceStatus.Unknown, ServiceStatusParser.Indicator(""));
    }

    [Fact]
    public void EveryCaseHasAShortLabelThatFitsTheDial()
    {
        Assert.Equal("OK", ServiceStatusText.Label(ServiceStatus.Operational));
        Assert.Equal("SLOW", ServiceStatusText.Label(ServiceStatus.Degraded));
        Assert.Equal("PARTIAL", ServiceStatusText.Label(ServiceStatus.PartialOutage));
        Assert.Equal("DOWN", ServiceStatusText.Label(ServiceStatus.MajorOutage));
        Assert.Equal("MAINT", ServiceStatusText.Label(ServiceStatus.Maintenance));
        Assert.Equal("—", ServiceStatusText.Label(ServiceStatus.Unknown));
    }

    [Fact]
    public void InLineModeTheLineSpeaksEvenWhenTheServiceIsFine()
    {
        // A silent line reads as a lost dial — that already happened once.
        Assert.Equal("service operational", ServiceStatusText.Line(ServiceStatus.Operational, StatusMode.Line));
        Assert.Equal("service degraded", ServiceStatusText.Line(ServiceStatus.Degraded, StatusMode.Line));
    }

    [Fact]
    public void WithADialInEveryBlockTheLineOnlyReportsTrouble()
    {
        Assert.Null(ServiceStatusText.Line(ServiceStatus.Operational, StatusMode.Cell));
        Assert.Equal("service major outage", ServiceStatusText.Line(ServiceStatus.MajorOutage, StatusMode.Cell));
    }

    [Fact]
    public void TheLineSpellsOutWhatTheDialAbbreviates()
    {
        // Label is cropped to fit inside a 68 pt dial; the line has room for
        // the word, and "service SLOW" is not a sentence.
        Assert.Equal("under maintenance", ServiceStatusText.LineLabel(ServiceStatus.Maintenance));
        Assert.Equal("state unknown", ServiceStatusText.LineLabel(ServiceStatus.Unknown));
    }
}

public class StatusDecoderTests
{
    private const string Payload = """
    {
      "status": { "indicator": "none", "description": "All Systems Operational" },
      "components": [
        { "name": "claude.ai", "status": "operational" },
        { "name": "Claude Code", "status": "degraded_performance" },
        { "name": "Claude Cowork", "status": "operational" }
      ]
    }
    """;

    [Fact]
    public void TheClaudeCodeComponentWinsOverThePageRollUp() =>
        Assert.Equal(ServiceStatus.Degraded, StatusDecoder.Status(Payload));

    [Fact]
    public void PrefersClaudeCodeComponent()
    {
        var status = StatusDecoder.Status("""
        {
          "components": [
            { "name": "Other", "status": "major_outage" },
            { "name": "Claude Code", "status": "operational" }
          ],
          "status": { "indicator": "critical" }
        }
        """);
        Assert.Equal(ServiceStatus.Operational, status);
    }

    [Fact]
    public void ThePageIndicatorIsTheFallbackWhenTheComponentIsAbsent()
    {
        const string json = """
        {
          "status": { "indicator": "critical" },
          "components": [ { "name": "claude.ai", "status": "operational" } ]
        }
        """;
        Assert.Equal(ServiceStatus.MajorOutage, StatusDecoder.Status(json));
    }

    [Fact]
    public void ABodyWithNeitherComponentNorIndicatorIsMalformed() =>
        Assert.Equal(UsageError.MalformedResponse,
            Assert.Throws<UsageException>(() => StatusDecoder.Status("""{"components": []}""")).Error);

    [Fact]
    public void ANonJsonBodyIsMalformed() =>
        Assert.Equal(UsageError.MalformedResponse,
            Assert.Throws<UsageException>(() => StatusDecoder.Status("<html>nope</html>")).Error);
}
